### PR TITLE
feature: streaming server of cri manager

### PR DESCRIPTION
**Pull request by YaoZengzeng**
Signed-off-by: YaoZengzeng <yaozengzeng@zju.edu.cn>

**1.Describe what this PR did**

**2.Does this pull request fix one issue?** 

fixes part of #420 

**3.Describe how you did it**

**4.Describe how to verify it**

```
// Now we must ensure the sandbox image has already exists.
[root@pouch-test cri-tools]# pouch pull k8s.gcr.io/pause-amd64:3.0
[root@pouch-test cri-tools]# cat sandbox-config.json
{
    "metadata": {
        "name": "nginx-sandbox",
        "namespace": "default",
        "attempt": 1,
        "uid": "hdishd83djaidwnduwk28bcsb"
    },
    "linux": {
    }
}
[root@pouch-test cri-tools]# crictl runs sandbox-config.json
008cc69c9c9c8e9c4d78aa8d6c528e21ab26affdc64748f631fb5dcb59963216
[root@pouch-test cri-tools]# cat container-config.json 
{
  "metadata": {
      "name": "busybox"
  },
  "image":{
      "image": "docker.io/library/busybox:latest"
  },
  "command": [
      "top"
  ],
  "linux": {
  }
}
[root@pouch-test cri-tools]# crictl create 008cc69c9c9c8e9c4d78aa8d6c528e21ab26affdc64748f631fb5dcb59963216 container-config.json sandbox-config.json
3530bbb414ccce09ba52fbdf781df53b881661720f05d92edf8820280abf064d
[root@pouch-test cri-tools]# crictl start 3530bbb414ccce09ba52fbdf781df53b881661720f05d92edf8820280abf064d
[root@pouch-test cri-tools]# pouch ps
Name                                                            ID       Status    Image                              Runtime   Created
k8s_busybox_nginx-sandbox_default_hdishd83djaidwnduwk28bcsb_0   3530bb   running   docker.io/library/busybox:latest   runc      4 seconds ago
k8s_POD_nginx-sandbox_default_hdishd83djaidwnduwk28bcsb_1       008cc6   running   k8s.gcr.io/pause-amd64:3.0         runc      1 minute ago
[root@pouch-test cri-tools]# crictl exec 3530bbb414ccce09ba52fbdf781df53b881661720f05d92edf8820280abf064d ls
FATA[0000] execing command in container failed: Internal error occurred: error executing command in container: streamRuntime's Exec Not Implemented Yet
```
**5.Special notes for reviews**

### CODE REVIEW

File: daemon/config/config.go [hunk 1]
```diff
@@ -20,6 +20,13 @@ type Config struct {
 	// ListenCRI is the listening address which serves CRI.
 	ListenCRI string
 
+	// StreamServerAddress is the ip address streaming server of CRI is listening on.
+	// The default host interface is used if not specified.
+	StreamServerAddress string
```
Comment (allencloud) [inline daemon/config/config.go]: Could we encapsulate a CRI config struct to keep all configs about CRI implementation?

Comment (YaoZengzeng) [reply daemon/config/config.go]: @allencloud  Yes, it is a good idea :)

**Comment by WIZARD-CXY** [comment]
too much code vendoring I think.

File: daemon/mgr/cri_streaming.go [hunk 2]
```diff
@@ -0,0 +1,50 @@
+package mgr
+
+import (
+	"fmt"
+	"io"
+	"net"
+
+	k8snet "k8s.io/apimachinery/pkg/util/net"
```
Comment (WIZARD-CXY) [inline daemon/mgr/cri_streaming.go]: Can we  only import "k8s.io/apimachinery/pkg/util/net" , not the whole apimachinery pkg?

File: daemon/mgr/cri_streaming.go [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+	"k8s.io/client-go/tools/remotecommand"
```
Comment (WIZARD-CXY) [inline daemon/mgr/cri_streaming.go]: can we only import just k8s.io/client-go/tools/remotecommand? not the whole pkg?

File: daemon/mgr/cri_streaming.go [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+	"k8s.io/kubernetes/pkg/kubelet/server/streaming"
+)
+
+func newStreamServer(ctrMgr ContainerMgr, address, port string) (streaming.Server, error) {
+	if address == "" {
+		// If address is "", we would use the host's default interface.
+		addr, err := k8snet.ChooseBindAddress(nil)
```
Comment (WIZARD-CXY) [inline daemon/mgr/cri_streaming.go]: can we just implement a ChooseBindAddress method? not import a whole apimachinery pkg?

File: daemon/mgr/cri_streaming.go [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+	return &streamRuntime{containerMgr: ctrMgr}
+}
+
+// Exec executes a command inside the container.
+func (s *streamRuntime) Exec(containerID string, cmd []string, stdin io.Reader, stdout, stderr io.WriteCloser, tty bool, resize <-chan remotecommand.TerminalSize) error {
+	return fmt.Errorf("streamRuntime's Exec Not Implemented Yet")
+}
+
+// Attach attaches to a running container.
+func (s *streamRuntime) Attach(containerID string, in io.Reader, out, err io.WriteCloser, tty bool, resize <-chan remotecommand.TerminalSize) error {
```
Comment (WIZARD-CXY) [inline daemon/mgr/cri_streaming.go]: the same as the ChooseBindAddress.

**Comment by YaoZengzeng** [comment]
@WIZARD-CXY I'll try to make dependency as less as possible :)

**Comment by YaoZengzeng** [comment]
Close this PR, the conflict notification is so annoying 😄 

I'm trying to implement the streaming server by myself instead of vendor so many dependencies :)

**Pull request closed without merging**